datatree/datatree.py

```diff
@@ -341,7 +341,22 @@ def __init__(
             attrs=ds._attrs,
             encoding=ds._encoding,
         )
-        self._close = ds._close
+        self._close = None if data is None else data._close
+
+    def close(self) -> None:
+        """Release any resources linked to this object."""
+        if self._close is not None:
+            self._close()
+        self._close = None
+
+        for child in self._children.values():
+            child.close()
```

Comment:
If you prefer not to do this recursively, you could instead use `.subtree`, i.e.

```python
def close(self):
    """Close any resources linked to any node in this subtree"""
    for node in self.subtree:
        node._close()  # close nodes one by one
```